src/gradualizer_lib.erl

```diff
@@ -104,10 +104,20 @@ get_type_definition({remote_type, _Anno, [{atom, _, Module}, {atom, _, Name}, Ar
     gradualizer_db:get_type(Module, Name, Args);
 get_type_definition({user_type, Anno, Name, Args}, TEnv, Opts) ->
     %% Let's check if the type is a known remote type.
-    case typelib:get_module_from_annotation(Anno) of
-        {ok, Module} ->
-            gradualizer_db:get_type(Module, Name, Args);
-        none ->
+    KnownTy = case typelib:get_module_from_annotation(Anno) of
+                  {ok, Module} ->
+                      gradualizer_db:get_type(Module, Name, Args);
+                  none ->
+                      no_module
+              end,
+    case KnownTy of
+        {ok, Ty} ->
+            %% gradualizer_db already did annotate_user_types
+            {ok, Ty};
+        opaque ->
+            opaque;
+        _ ->
+            %% _ = (no_module | not_found)
```

Comment:
I was running into a situation when a type from the currently checked module (i.e. not in the DB) was available in `TEnv`, but since the lookup to the DB was done first, the lookup failed, since we didn't try to consider `TEnv` at all. I think this is a bug.

Maybe the problem stemmed from the fact that the type had the module information present, and that's why it looked in the DB first, but OTOH if it was a remote record type, it had to have the module info (actually taken from the file/location annotation - `typelib:get_module_from_annotation`) to disambiguate records of the same names defined in different modules.

> We shouldn't fallback from remote type not found to lookup local type. That would be an error AFAICT.

If that's the case we have to find a better way to choose whether a lookup has to be done in the DB or TEnv than:
```
case typelib:get_module_from_annotation(Anno) of
```

OR we have to find another way to disambiguate between same-name records defined in different modules.

Reply:
Ahh, one more note on that - records, at least often, not sure if in 100% of the cases, are passed around as `{type,1,record,[{atom,1,rec} = _Name]}`, i.e. without the field information. This lookup is necessary to get the fields either from `TEnv` or the DB.

Reply:
Yes, the original idea was to use `typelib:get_module_from_annotation` to disambiguate between the same record name defined in different modules, or if no module annotation, the current module. What's wrong with that?

> Ahh

Are you saying that sometimes when we through away the field information we also throw away the annotation in record types? That's the bug which needs to be fixed then.

The fields are usually not present in a record type. Only the fields that have a "redifined" type are present, if any.

Reply:
> Are you saying that sometimes when we through away the field information we also throw away the annotation in record types?

No. We throw away the field info, but we leave the file info. Then the lookup is done to the DB, whereas the record type is not in the DB - it's in TEnv.

Reply:
Oops. But now that `remove_pos` is fixed for records again, this can't happen anymore, right?

Reply:
But even the fixed `remove_pos` doesn't throw away the file info, only the line info.

Reply:
Oh, you're right.

```Erlang
%% Removes all annotations from type, except filename in two cases: Filename is
%% kept for user-defined types and record types. Filename is used to
%% disambiguate between types with the same name from different modules.
%% Annotated types as in Name :: Type are also removed.
-spec remove_pos(type()) -> type().
```

Maybe we should remove the filename from local records and user types (maybe all annotations from all types), directly when we create TEnv. In this way, we'll only have filename for types coming from db.

Reply:
I've removed the "Fix leaky lookup in gradualizer_lib:get_type_definition/3" commit and I don't see the problem coming back yet. If I do, I'll try to write a corresponding test.